### ISSUE TITLE
MoveAndSlide / IsOnFloor and MoveAndCollide don't work in C# scripts

### ISSUE THREAD

**Issue by copygirl** [godotengine/godot#19898]
**Godot:** `3.1.dev.mono.custom_build.896e250`
**OS:** Arch Linux

### Issue description
I tried following the FPS tutorial and recreate the 3D movement in a from-scratch project using C# scripts but I quickly ran into the issue of `IsOnFloor()` always returning `false` no matter what I tried.

`MoveAndCollide` also never returns a collision even though the entity stops as it collides with the floor.

### Minimal reproduction project / Steps to reproduce
See the [`Player.cs`](https://github.com/copygirl/Immersion/blob/5cae93c7d01d81146b7d26c56ff33b316024b48b/Player.cs#L44) file of [my project](https://github.com/copygirl/Immersion).
Simply... let the player collide with the ground and see that the printed values stay false.

**edit:** It was suggested for me to increase the gravity, safe margin and check `IsOnFloor()` after `MoveAndSlide()`, so [I did so](https://github.com/copygirl/Immersion/commit/f31a78657083ef1db6296826f4fc90f2a24ca48a), but still no luck.

**edit2:** I also tried [re-implementing `MoveAndSlide`](https://github.com/copygirl/Immersion/blob/bbdc906f1db9854eb5d0c1ffcfed00d2aed71260/Player.cs#L98) myself.

**Comment by CarlGustavAlbertDwarfsteinYung** [comment]
Make sure your Floor Normal vector is actually pointing Up. Without specifying a floor vector the game won't know which way is up or down because by default is Vector3(0,0,0)  

![image](https://user-images.githubusercontent.com/36826319/42144225-0130f0d4-7d88-11e8-850e-7c5ce9b81b00.png)
It needs to be Vector3(0,1,0) pointing UP

But even after you do this it will still not work properly. This is because move_and_slide attemps a number of slides and some of them will bounce off the ground which means your is_on_floor() is gonna look like true false true false true false ... alternating. 

I suggest implementing your own method.

**Comment by copygirl** [comment]
I've tried `new Vector3(0, 1, 0)`, `Vector3.Up` (current) and `Vector3.Down`. The `IsOnFloor()` method at no point returns true, it appears. The FPS tutorial project doesn't have this problem. At this point I'm wondering if somehow the `IsOnFloor` binding isn't working at all.

**Comment by int010h** [comment]
Try to call IsOnFloor after MoveAndSlide. Also increacing gravity acceleration and KinematicBody's safe margin seems to have effect on IsOnFloor behavior, you can play around and tweak. But still, none of this worked for me in case of colliding walls while standing on the floor, IsOnFloor seems to be flickering from true to false randomly. So I ended up replacing it with Raycast node until it's fixed, seems to be working well for now.

**Comment by copygirl** [comment]
@int010h I tried what you suggested, but [the changes](https://github.com/copygirl/Immersion/commit/f31a78657083ef1db6296826f4fc90f2a24ca48a) still did not have an effect on the outcome. I will likely try to work around the issue like you recommended, though it was sort of important for me to report it.

**Comment by jmf** [comment]
This seems to be a general issue with IsOnFloor() in the mono build. It doesn't appear to work at all. I was able to replicate this bug with 3.0.4-mono from the website. This bug doesn't seem to affect the is_on_floor() function in gdscript.

**Comment by copygirl** [comment]
I tinkered around a little bit, re-implementing `MoveAndSlide` in C# and found out that `MoveAndCollide` [never even returns a collision](https://github.com/copygirl/Immersion/blob/bbdc906f1db9854eb5d0c1ffcfed00d2aed71260/Player.cs#L109-L111), yet stops the player from moving any further.

I might look into doing the same for `MoveAndCollide` but I'm not sure yet as I haven't taken a look at it; it's probably a bit more complicated than the other method.

**Comment by RySchmitt** [comment]
I think this caused by #5910. As a new user of Godot, I ran into this exact issue in C#, but I was able to fix it by changing the collision shape of the floor from a PlaneShape to a CubeShape. I downloaded @copygirl 's project and the same change, with a couple tweaks, fixed the issue. The FPS tutorial in GodotScript also breaks if the floor is turned into a plane. It seems that the problem has nothing to do with the C# implementation.

I believe this issue should be closed.
